test-runner/src/main/scala/scala/scalanative/testinterface/ComRunner.scala

```diff
@@ -68,24 +70,32 @@ class ComRunner(bin: File,
   /** Wait for a message to arrive from the distant program. */
   def receive(timeout: Duration = Duration.Inf): Message =
     synchronized {
-      in.mark(Int.MaxValue)
       val savedSoTimeout = socket.getSoTimeout()
+
       try {
-        val deadLineMs = if (timeout.isFinite()) timeout.toMillis else 0L
-        socket.setSoTimeout((deadLineMs min Int.MaxValue).toInt)
+        // Java sockets only handle non-negative Int timeouts.
+        // Silently truncate timouts greater than Int.MaxValue.
+        val deadLineMs = if (!timeout.isFinite()) {
+          0
+        } else {
+          timeout.toMillis.toInt
+        }
+
+        if (socketCurrentSoTimeout != deadLineMs) {
+          socket.setSoTimeout(deadLineMs)
+          socketCurrentSoTimeout = deadLineMs
+        }
 
-        val result =
+        var result: Message = new Log(0, "Dummy", None, Level.Info)
+
+        while (result.isInstanceOf[Log]) {
           SerializedInputStream.next(in)(_.readMessage()) match {
-            case logMsg: Log =>
-              log(logMsg)
-              receive(timeout)
-            case other =>
-              other
+            case logMsg: Log => log(logMsg)
+            case other       => result = other
```

Comment:
If you want to use a loop, just use a `while (true)` and do a `return other` here. Looping while the result is a `Log` is really weird, and doesn't help readability IMO. The problem with the `while (true)` is that you'll need a `throw new AssertionError("unreachable")` after the while loop to convince scalac that you don't need a `Message` after the loop, so that's not great either.

IMO the most readable version would a tail-recursive local `loop()` function in this case. Something like:
```scala
@tailrec
def loop(): Message = {
  SerializedInputStream.next(in)(_.readMessage()) match {
    case logMsg: Log =>
      log(logMsg)
      loop()
    case other =>
      other
  }
}

loop()
```
I agree that recursively calling `receive()` itself was not great, because of the actual recursion, which was even within the try/catch, leading to really weird traces. A local tail-rec def won't have those issues.